### PR TITLE
*: Tag io type for foreground and snpashot

### PR DESCRIPTION
**Pull request by Connor1996**
Signed-off-by: Connor <zbk602423539@gmail.com>

### What problem does this PR solve?

Issue Number: part of #9156 
Problem Summary: 
Tag io type for foreground read/write and snapshot of replication/loadbalance, so io rate limiter can use it to throttle different tasks

### What is changed and how it works?

What's Changed:
- tag `ForegroundRead` for storage/coprocessor readpool
- tag `ForegroundWrite` for scheduler worker and batch system of raftstore/apply
- Add a field `for_balance` for snapshot meta https://github.com/pingcap/kvproto/pull/707, so the receiver can know whether the snapshot is triggered by balance.
- use `matched_index==0` to decide whether the snapshot is triggered by balance or not, and tag `LoadBalance` or `Replication` on the generation and receiving of snapshots. 

### Related changes

https://github.com/pingcap/kvproto/pull/707

### Check List 

- Manual test (add detailed scripts or steps below)

### Release note 

- None

### CODE REVIEW

**Comment by Connor1996** [comment]
PTAL @tabokie

File: components/raftstore/src/store/worker/region.rs [hunk 1]
```diff
@@ -277,16 +281,23 @@ where
         last_applied_state: RaftApplyState,
         kv_snap: EK::Snapshot,
         notifier: SyncSender<RaftSnapshot>,
+        for_balance: bool,
     ) {
         SNAP_COUNTER.generate.all.inc();
         let start = tikv_util::time::Instant::now();
+        let _ = WithIOType::new(if for_balance {
```
Comment (tabokie) [inline components/raftstore/src/store/worker/region.rs:288]: I made a similar mistake recently. `_` will actually be dropped immediately. Use `_io_type_guard` instead.

File: src/server/snap.rs [hunk 2]
```diff
@@ -199,8 +201,16 @@ impl RecvSnapContext {
             Err(e) => return Err(box_err!("failed to create snap key: {:?}", e)),
         };
 
+        let data = meta.get_message().get_snapshot().get_data();
+        let mut snapshot = RaftSnapshotData::default();
+        snapshot.merge_from_bytes(data)?;
```
Comment (tabokie) [inline src/server/snap.rs:206]: Is this a decoding procedure? How much overhead?

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -2756,16 +2756,23 @@ pub struct CatchUpLogs {
...
 impl GenSnapTask {
     pub fn new(region_id: u64, snap_notifier: SyncSender<RaftSnapshot>) -> GenSnapTask {
         GenSnapTask {
             region_id,
             snap_notifier,
+            for_balance: false,
         }
     }
 
+    pub fn mark_for_balance(&mut self) {
```
Comment (tabokie) [inline components/raftstore/src/store/fsm/apply.rs:2772]: stick to `set_for_balance`?

**Review by tabokie — COMMENTED** [review]
Looks good generally. Should wait on #9340 to add more tests.

**Review by tabokie — COMMENTED** [review]
-

File: components/raftstore/src/store/peer.rs [hunk 4]
```diff
@@ -1574,7 +1574,12 @@ where
 
         if !self.raft_group.has_ready() {
             // Generating snapshot task won't set ready for raft group.
-            if let Some(gen_task) = self.mut_store().take_gen_snap_task() {
+            if let Some(mut gen_task) = self.mut_store().take_gen_snap_task() {
+                if let Some(progress) = self.raft_group.status().progress {
+                    if progress.iter().any(|(_, pr)| pr.matched == 0) {
```
Comment (yiwu-arbug) [inline components/raftstore/src/store/peer.rs:1579]: not familiar with the code here. what does `pr.matched == 0` mean?

File: components/raftstore/src/store/worker/region.rs (hunk 1 above)
Comment (yiwu-arbug) [reply components/raftstore/src/store/worker/region.rs:288]: TIL

**Review by yiwu-arbug — COMMENTED** [review]
Add some UT to test various cases of IO metrics?

File: components/raftstore/src/store/peer.rs (hunk 4 above)
Comment (Connor1996) [reply components/raftstore/src/store/peer.rs:1579]: It relies on the fact that when a peer is uninitialized (e.g. created by load balance), the last index of the peer is 0 which makes the matched index to be 0.

Comment (Connor1996) [reply components/raftstore/src/store/peer.rs:1579]: PTAL @hicqu

Comment (yiwu-arbug) [reply components/raftstore/src/store/peer.rs:1579]: Add comment for it?

**Review by yiwu-arbug — APPROVED** [review]
LGTM

File: tests/integrations/raftstore/test_snap.rs [hunk 5]
```diff
@@ -481,3 +483,39 @@ fn test_request_snapshot_apply_repeatedly() {
...
+    assert_ne!(recorder.fetch(IOType::Replication, IOOp::Write), 0);
+    
+    pd_client.must_remove_peer(1, new_peer(2, 2));
+    assert_eq!(recorder.fetch(IOType::LoadBalance, IOOp::Read), 0);
+    assert_eq!(recorder.fetch(IOType::LoadBalance, IOOp::Write), 0);
+    pd_client.must_add_peer(1, new_peer(2, 2));
+    must_get_equal(&cluster.get_engine(2), b"k2", b"v2");
+    assert_ne!(recorder.fetch(IOType::LoadBalance, IOOp::Read), 0);
+    assert_ne!(recorder.fetch(IOType::LoadBalance, IOOp::Write), 0);
+}
```
Comment (tabokie) [inline tests/integrations/raftstore/test_snap.rs:521]: \n

**Review by tabokie — APPROVED** [review]

**Review by yiwu-arbug — APPROVED** [review]

**Pull request merged**